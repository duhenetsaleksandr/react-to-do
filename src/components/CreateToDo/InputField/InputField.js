import React, { useState, useRef, useEffect } from 'react';
import { useIntl } from 'react-intl'

const initialState = '';

const InputField = ({ editTodo, onCreated, handlerBlurInput }) => {
    const refInputTodo = useRef(null);
    const [ inputValue, setInputValue ] = useState(initialState);
    const intl = useIntl();

    useEffect(() => {
        focusInput();
        const currentState = editTodo.title ? editTodo.title : initialState;
        setInputValue(currentState);
    }, [editTodo, onCreated]);

    const handlerChangeInput = ({ target: { value } }) => {
        setInputValue(value);
    };

    const focusInput = () => {
        refInputTodo.current.focus();
        refInputTodo.current.selectionStart = refInputTodo.current.value.length;
        refInputTodo.current.selectionEnd = refInputTodo.current.value.length;
    };

    const handlerKeyDown = ({ code }) => {
        if (code === 'Enter') handlerBlurInput(inputValue);
    };

    return (
        <input
            type="text"
            placeholder={intl.formatMessage({ id: 'placeholder' })}
            value={inputValue}
            onChange={handlerChangeInput}
            onBlur={handlerBlurInput.bind(null, inputValue)}
            onKeyDown={handlerKeyDown}
            ref={refInputTodo}
        />
    );
}

export default InputField;
