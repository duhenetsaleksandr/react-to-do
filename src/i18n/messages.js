import { LOCALES } from './locales'

export const messages = {
    [LOCALES.ENGLISH]: {
        copyright: '©Copyright. All rights deserved.',
        button_add: 'Add task',
        button_edit: 'Edit task',
        placeholder: 'Enter new todo',
        not_todos: 'Not found todos!',
    },
    [LOCALES.RUSSIAN]: {
        copyright: '©Копирайт. Все права защищены.',
        button_add: 'Добавить',
        button_edit: 'Редактировать',
        placeholder: 'Введите новую задачу',
        not_todos: 'Задачи не найдены!',
    },
};
