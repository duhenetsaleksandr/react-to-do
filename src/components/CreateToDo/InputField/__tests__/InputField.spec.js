import React from 'react';
import { IntlProvider } from 'react-intl';
import { act } from "@testing-library/react";
import { messages } from "i18n/messages";
import { LOCALES } from "i18n/locales";
import InputField from '../InputField';

const props = {
    editedTodo: {},
    flagCreated: true,
    onBlurInput: jest.fn(),
};

describe('should render InputField component', () => {
    let component;
    beforeEach(() => {
        component = mount(<InputField {...props} />, {
            wrappingComponent: IntlProvider,
            wrappingComponentProps: {
                messages: messages[LOCALES.ENGLISH],
                locale: LOCALES.ENGLISH,
                defaultLocale: LOCALES.ENGLISH,
            },
        });
    });

    it('should call handlerBlurInput props after keydown Enter', () => {
        act(() => {
            component.find('input').getElement().props.onChange({ target: { value: 'example test' } });
        });
        component.update();
        act(() => {
            component.find('input').getElement().props.onKeyDown({ code: 'Enter' });
        });
        component.update();
        expect(props.onBlurInput).toHaveBeenCalledWith('example test');
    });

    it('should not call handlerBlurInput props after keydown not Enter', () => {
        act(() => {
            component.find('input').getElement().props.onChange({ target: { value: 'example test' } });
        });
        component.update();
        act(() => {
            component.find('input').getElement().props.onKeyDown({ code: 'Escape' });
        });
        component.update();
        expect(props.onBlurInput).not.toHaveBeenCalled();
    });

    it('should change input value in state with onChange event', () => {
        act(() => {
            component.find('input[type="text"]').getElement().props.onChange({ target: { value: 1 } });
        });
        component.update();
        expect(component.find('input[type="text"]').getElement().props.value).toBe(1);
    });

    it('should change input value in state after update props', () => {
        act(() => {
            component.setProps({
                editedTodo: {
                    title: 'For test'
                },
                flagCreated: true
            });
        });
        component.update();
        expect(component.find('input[type="text"]').getElement().props.value).toBe('For test');
    });

    it('should render InputField component snapshot', () => {
        const component = shallow(
            <IntlProvider messages={messages[LOCALES.ENGLISH]} locale={LOCALES.ENGLISH} defaultLocale={LOCALES.ENGLISH}>
                <InputField {...props} />
            </IntlProvider>
        );
        expect(component).toMatchSnapshot();
    });
});
