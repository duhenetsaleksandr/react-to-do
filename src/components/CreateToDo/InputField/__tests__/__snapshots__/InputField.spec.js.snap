// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render InputField component should render InputField component snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en-US",
      "defaultRichTextElements": undefined,
      "fallbackOnEmptyString": true,
      "formatDate": [Function],
      "formatDateTimeRange": [Function],
      "formatDateToParts": [Function],
      "formatDisplayName": [Function],
      "formatList": [Function],
      "formatListToParts": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatNumberToParts": [Function],
      "formatPlural": [Function],
      "formatRelativeTime": [Function],
      "formatTime": [Function],
      "formatTimeToParts": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getDisplayNames": [Function],
        "getListFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralRules": [Function],
        "getRelativeTimeFormat": [Function],
      },
      "locale": "en-US",
      "messages": Object {
        "button_add": "Add task",
        "button_edit": "Edit task",
        "copyright": "Â©Copyright. All rights deserved.",
        "placeholder": "Enter new todo",
      },
      "onError": [Function],
      "textComponent": Symbol(react.fragment),
      "timeZone": undefined,
      "wrapRichTextChunksInFragment": undefined,
    }
  }
>
  <InputField
    editTodo={Object {}}
    handlerBlurInput={[MockFunction]}
    onCreated={true}
  />
</ContextProvider>
`;
